### ISSUE TITLE
Your hardware + GPU setup for training and average time to complete training

### ISSUE THREAD

**Issue by subzerofun** [zsdonghao/SRGAN#3]
First off, **thanks for publishing the code**! 
I've looked at a lot of SRGAN implementations on github but yours is the first with an understandable code structure for Tensorflow beginners.

I just wondered how your hardware setup looked like: CPU model, which GPU – and how many?

I'm using macOS 10.12.5 with an i7 6700k and two old crappy GPUs (for now, my GTX 1080 Ti arrives on Thursday! 🎉). Images get loaded from SSD (Read: ~600 MB/s).

| Model | GPU Memory    | CUDA Cores | CUDA V.  | GPU Max Clock rate | Mem Clock rate | Mem Bus Width: |
|-------------|-----------|------------|------------------|--------------------|-------------------|-------------------|
| GTX 780     | 6144 MB   | 2304       | 3.5              | 902 MHz            | 3004 Mhz          | 384-bit           |
| GTX 770     | 2048 MB   | 1536       | 3.0              | 1110 MHz           | 3505 Mhz          | 256-bit           |
| SOON: | | | | | | | |
| GTX 1080 Ti | 11.264 MB | 3584       | 6.1              | ~1700 MHz          | 11010 Mhz         | 352 bit           |

So for now i just have the two 700 series cards available. Since Tensorflow 1.2 doesn't offer GPU support for macOS i had to revert to 1.1.

**Is it normal that i have to wait around 10min until the training actually starts?** 
It takes 10min from entering the cmd until i see the Epochs counting down:
`Epoch [60/2000]  199 time: 2.3310s, d_loss: 0.13401291 g_loss: 0.03180185 (mse: 0.015155 vgg: 0.014024 adv: 0.002623)`

I have completed the first 100 Epochs for SRGAN init, and am now @ Epoch Nr. 60/2000 of the actual training. But it would take me ~11 days to get to step 2000! 

Do you think the 1080 Ti could cut the training time down to a bearable amount?

**How long did it take for you to complete the 2000 Epochs?**

And one more (so i don't have to open a separate issue): 
**When i'm done with training can i input an image that's not from the training set to upscale it? Could you please explain what i would need to change/add to the code?**

**Comment by zsdonghao** [comment]
Hi, I am using Titan X Pascal, 32GB RAM. I think 1080 Ti should be ok ~~

It is unnormal to wait 10 mins to let it start to train. May be the compile time of SubpixelConv is very long in your machine? It may cause by CPU.

It takes me 2 days to complete 2000 epochs when I use DIV2K dataset.

After you train the model, you can input any image to upscale it, check the `def evalution()` in `main.py`

**Comment by subzerofun** [comment]
Thanks for the quick answer! Can't figure out the reason for the long setup time...

The first steps go by pretty fast:

```
  [TL] FlattenLayer VGG19/flatten: 25088
  [TL] DenseLayer  VGG19/fc6: 4096 relu
  [TL] DenseLayer  VGG19/fc7: 4096 relu
  [TL] DenseLayer  VGG19/fc8: 1000 identity
build model finished: **0.553778s**
build model started
```

```
  [TL] FlattenLayer VGG19/flatten: 25088
  [TL] DenseLayer  VGG19/fc6: 4096 relu
  [TL] DenseLayer  VGG19/fc7: 4096 relu
  [TL] DenseLayer  VGG19/fc8: 1000 identity
build model finished: **0.136154s**
```

But then it freezes for a few minutes at

```
  [TL] SubpixelConv2d  SRGAN_g/pixelshufflerx2/2: scale: 2 n_out_channel: 64 act: relu
  [TL] Conv2dLayer SRGAN_g/out: shape:[1, 1, 64, 3] strides:[1, 1, 1, 1] pad:SAME act:tanh
  [*] geting variables with SRGAN_g
```
and then again for a few minutes at

```
got  31: SRGAN_d/res/bn3/gamma:0   (512,)
got  32: SRGAN_d/ho/dense/W:0   (18432, 1)
got  33: SRGAN_d/ho/dense/b:0   (1,)
```

Nevertheless, when it starts the speed is OK (as good as it can be with my 700-series cards). I'm really excited to swap out the 770 on Thursday and see how well the 1080 Ti will do!

**Comment by Andreababy** [comment]
Can you download the code?

**Comment by Andreababy** [comment]
i cannot download it sucessful. Do you know the reason? Thank you.

**Comment by subzerofun** [comment]
What code do you mean specifically? The whole repository? Just enter terminal, cd into a folder where you store your projects and then execute:
`git clone https://github.com/zsdonghao/SRGAN.git`

The output should be:
```
Cloning into 'SRGAN'...
remote: Counting objects: 228, done.
remote: Compressing objects: 100% (13/13), done.
Receiving objects: 100% (228/228), 117.82 MiB | 2.69 MiB/s, done.
Resolving deltas: 100% (111/111), done.
```

**Comment by rlayne** [comment]
@subzerofun I have the same concern as you.  The model takes quite some time to actually get down to training.  I suspect it is a CPU thing, since I'm running this on my home PC which packs significantly less CPU cores than a server.  Did you find a way to speed things up?

**Comment by subzerofun** [comment]
No, unfortunately not... The first time i tried my CPU speed was reaching 3,8 Ghz with Turbo Boost. I forgot to disable Speed Stepping in the BIOS and went back, changed the CPU speed to be at 4.2 Ghz at all times, turned off Speed Stepping and now the training setup takes 2-3 minutes less.

I was already at epoch Nr. 570 of 2000 and then my SSDs disks space was full... Argh. I lost all my progress because the .npz file with the weights is now corrupt...

So i have to start again from epoch 60... two days lost for nothing.

**So make sure you backup your .npz files every now and then and note the epoch you were at if you pause training so that you don't lose your progress like i did!**

I will try to compile Tensorflow from source since i always get messages
`tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.`
`W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.`

I don't know if that will speed up things, but i will report if it does.

Another thing which may help is creating a RAM disk, copying all the repo files there and try again.

In the CUDA documentation there is an option to set your cache folder, which you could also point to the RAM disk.

But since it's probably a CPU related issue i don't know how much this is gonna affect things.

**Comment by zsdonghao** [comment]
`SubpixelConv2d` takes long time to compile, but it is a important part in the original paper.
However, if you like, you can use resize-convolution i.e. use `UpSamplingLayer` and `Conv2d` instead of `SubpixelConv2d` as `model.py --> SRGAN_g2()` shows.

check this link for resize-conv http://distill.pub/2016/deconv-checkerboard/

**Comment by subzerofun** [comment]
@zsdonghao Will this affect the output quality? I'd rather wait a few minutes and have a better model when the actual training takes a few days than using a shortcut and getting worse results.

BTW sorry that i mention this in this thread, but could you please take a look at this issue? It won't take up much your time! [https://github.com/zsdonghao/Unsup-Im2Im/issues/9](https://github.com/zsdonghao/Unsup-Im2Im/issues/9)

**Comment by Andreababy** [comment]
I cannot download the pretrained VGG19 model.

**Comment by rlayne** [comment]
Hmm. Probably need to figure a way of caching that part of the program then, if possible.  I'm not too advanced on these matters yet to know if it's possible.  On a 16-core machine perhaps it's less of an issue.

Good advice, cheers.  

Yes, RAM disks are a standard approach to speeding things up if you have enough memory and the bottleneck is feeding the network.  I'd just hold the data there, I don't expect putting the code itself there to help speed much (particularly on 'nix systems after accessing the files at least once, I think the system will automagically put them in there anyway).

**Comment by subzerofun** [comment]
@Andreababy If you can't downlod from mega.nz try one of these:
* 1) Check if a plugin/extension is affecting the site or try another browser
* 2) Use jDownloader, a download helper & organiser tool: [http://jdownloader.org/jdownloader2](http://jdownloader.org/jdownloader2). Install jDownoader, rightclick and copy the [link](https://mega.nz/#!xZ8glS6J!MAnE91ND_WyfZ_8mvkuSa2YcA7q-1ehfSm-Q1fxOvvs) and as soon as its in the clipboard  jDownloader will add it in your "LinkGrabber" tab. Rightclick on the mega.nz link (under the LinkGrabber tab), select "Start Downloads". It should now be under the "Downloads" tab. The file will be in your standard OS downloads directory.
* 3) Download it here (only online till 16th of July 2017!): [https://we.tl/8vgsapn2qN](https://we.tl/8vgsapn2qN)

**Comment by Andreababy** [comment]
thank you very much发自我的华为手机-------- 原始邮件 --------主题：Re: [zsdonghao/SRGAN] Your hardware + GPU setup for training and average time to complete training (#3)发件人：subzerofun 收件人：zsdonghao/SRGAN 抄送：Andreababy <2018171324@qq.com>,Mention @Andreababy If you can't downlod from mega.nz try one of these:



Check if a plugin/extension is affecting the site or try another browser




Use jDownloader, a download helper & organiser tool: http://jdownloader.org/jdownloader2. Install jDownoader, rightclick and copy the link and as soon as its in the clipboard  jDownloader will add it in your "LinkGrabber" tab. Rightclick on the mega.nz link (under the LinkGrabber tab), select "Start Downloads". It should now be under the "Downloads" tab. The file will be in your standard OS downloads directory.




Download it here (only online till 16th of July 2017!): https://we.tl/8vgsapn2qN




—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/zsdonghao/SRGAN","title":"zsdonghao/SRGAN","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/zsdonghao/SRGAN"}},"updates":{"snippets":[{"icon":"PERSON","message":"@subzerofun in #3: @Andreababy If you can't downlod from mega.nz try one of these:\r\n* 1) Check if a plugin/extension is affecting the site or try another browser\r\n* 2) Use jDownloader, a download helper \u0026 organiser tool: [http://jdownloader.org/jdownloader2](http://jdownloader.org/jdownloader2). Install jDownoader, rightclick and copy the [link](https://mega.nz/#!xZ8glS6J!MAnE91ND_WyfZ_8mvkuSa2YcA7q-1ehfSm-Q1fxOvvs) and as soon as its in the clipboard  jDownloader will add it in your \"LinkGrabber\" tab. Rightclick on the mega.nz link (under the LinkGrabber tab), select \"Start Downloads\". It should now be under the \"Downloads\" tab. The file will be in your standard OS downloads directory.\r\n* 3) Download it here (only online till 16th of July 2017!): [https://we.tl/8vgsapn2qN](https://we.tl/8vgsapn2qN)\r\n\r\n\r\n\r\n\r\n\r\n"}],"action":{"name":"View Issue","url":"https://github.com/zsdonghao/SRGAN/issues/3#issuecomment-313916493"}}}

**Comment by Andreababy** [comment]
I encounter trouble again , i cannot download the "DIV2K - bicubic downscaling x4 competition " dataset, can you give me a link? thank you very much

**Comment by subzerofun** [comment]
@Andreababy  I too had problems finding it – there was no download link on the competition site...

Here you go:
https://data.vision.ee.ethz.ch/cvl/DIV2K/validation_release/DIV2K_test_LR_bicubic_X4.zip
https://data.vision.ee.ethz.ch/cvl/DIV2K/DIV2K_train_HR.zip
https://data.vision.ee.ethz.ch/cvl/DIV2K/DIV2K_train_LR_bicubic_X4.zip
https://data.vision.ee.ethz.ch/cvl/DIV2K/validation_release/DIV2K_valid_HR.zip
https://data.vision.ee.ethz.ch/cvl/DIV2K/DIV2K_valid_LR_bicubic_X4.zip

What hardware do you plan to use for training?

**Comment by Andreababy** [comment]
GPU GTX108

**Comment by Andreababy** [comment]
In the five link, Which one should I download? Which one is the author using?

**Comment by rlayne** [comment]
@Andreababy You'll need the train and validation HR and LR links, at minimum.  You may as well get the test data too, but if I recall the code here only requires training and validation sets.  There is no HR set for the testing set.  If you need more data, simply scrape high resolution images from the web, and downsample them by a factor of 4 using bicubic resampling to create the corresponding "LR" set.

**Comment by Andreababy** [comment]
Thank you very much

**Comment by 42binwang** [comment]
@subzerofun Can you tell me how to run this SRGAN on multi-GPU system?

**Comment by Andreababy** [comment]
When you use GTX1080Ti, how long have you spent during the whole training? @subzerofun

**Comment by Andreababy** [comment]
Follow the readme, first,download the whole code,then download the VGG19 model as the readme shows,second, download the dataset,finally,run the code. @42binwang

**Comment by 42binwang** [comment]
@Andreababy This code seems only use one GPU.

**Comment by subzerofun** [comment]
@Andreababy my training is not finished yet, i'm at epoch 1700/2000. I will upload the weights once the training is complete. Should be ready on Friday.

**Comment by subzerofun** [comment]
@42binwang don't know about Multi-GPU, but afaik tensorflow automatically uses all available gpu devices. I'm currently running the training with two gpus, but it doesn't seem much faster compared to my tests with one gpu...

**Comment by Andreababy** [comment]
waiting your trained model. @subzerofun

**Comment by 42binwang** [comment]
@subzerofun You need to allocate the task on each GPUs. Tensorflow will use only the first GPU in default.

**Comment by Andreababy** [comment]
When i use my own picture, in the testing,i encountered this problem,what should i do? if i want to test my own picture,what should i    @subzerofun  @zsdonghao 


ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[1,1916,2400,256]
	 [[Node: SRGAN_g/n256s1/2/Conv2D = Conv2D[T=DT_FLOAT, data_format="NHWC", padding="SAME", strides=[1, 1, 1, 1], use_cudnn_on_gpu=true, _device="/job:localhost/replica:0/task:0/gpu:0"](SRGAN_g/SRGAN_g/pixelshufflerx2/1/Relu, SRGAN_g/n256s1/2/W_conv2d/read)]]
	 [[Node: SRGAN_g/out/Tanh/_349 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_835059_SRGAN_g/out/Tanh", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]

**Comment by subzerofun** [comment]
@Andreababy `OOM` means "Out of memory". You should try to free up more VRAM (close all other applications) or downscale the image. How much memory does your card have?

**Comment by Andreababy** [comment]
Can you tell me how to test my own picture？发自我的华为手机

**Comment by Andreababy** [comment]
i downscale the picture.it can test my own picture，but the testing quanlity is poor. it cannot raise the resolution.Can you tell me the reason？when i test my own picture， should i put my own picture in the test dataset？发自我的华为手机

**Comment by subzerofun** [comment]
look in main.py for this part:
```python
# valid_lr_img = get_imgs_fn('test.png', 'data2017/')  # if you want to test your own image
```
Uncomment the line. Then put in your own filename where it says "test.png" and maybe you need to change the "data2017/" folder too.

I haven't tried it, neither have i written any code for this project. So i'm just guessing.

To get good results you need the trained weights – if you upscale your images without a trained model, it won't work! I have trained weights at epoch 1870/2000 and you can [download them here.](https://wetransfer.com/downloads/aa53bfd9d6597f2a340c2a1e9da714e820170714234615/f92388af4d5d33a710d68529223eb43520170714234615/8db3f8)

If you have done some training yourself please save the npz files in your checkpoint folder somewhere else before you overwrite them. If you want to use my weights simply put the .npz files in your checkpoint folder. 

---

**Here are some results from my training:**

Image comparison slider:
Sample & SRGAN upscaled image: [https://codepen.io/subzerofun/full/qjvywo/](https://codepen.io/subzerofun/full/qjvywo/)
Original HR & SRGAN upscaled image: [https://codepen.io/subzerofun/full/MoxBNJ/](https://codepen.io/subzerofun/full/MoxBNJ/)

### Sample: 96 x 96 px upscaled to 400% (bilinear) for comparison:
![_train_sample_96-400 -bilinear](https://user-images.githubusercontent.com/17863119/28234463-1dc7361a-6900-11e7-80ed-10dab2738ab6.png)

### Original HR image:
![_train_sample_384](https://user-images.githubusercontent.com/17863119/28234473-3fbbabb6-6900-11e7-9b9d-474b6bb20798.png)

### Upscaled with SRGAN, training at epoch 1873/2000:
![1873 2000 _train](https://user-images.githubusercontent.com/17863119/28234486-5b74ee6c-6900-11e7-9744-84c71140ea12.png)

So i would say the result looks pretty good, but a lot of the fine details are very different compared to the original image.

**Comment by Andreababy** [comment]
i downscale the picture.it can test my own picture，but the testing quanlity is poor. it cannot raise the resolution.Can you tell me the reason？when i test my own picture， should i put my own picture in the test dataset？ @zsdonghao

**Comment by subzerofun** [comment]
@Andreababy Hey – i've already taken my time to answer your question...

1. Download my weights from [here](https://wetransfer.com/downloads/aa53bfd9d6597f2a340c2a1e9da714e820170714234615/f92388af4d5d33a710d68529223eb43520170714234615/8db3f8)
2. Put them in your `checkpoint` folder (move your own npz files somewhere else, don't overwrite them!)
3. Change the filename (your image!) and output folder in this line of main.py:
 ```# valid_lr_img = get_imgs_fn('test.png', 'data2017/')  # if you want to test your own image```. Uncomment the line (remove the #).
4. Run the code with `python main.py --mode=evaluate` 
5. Try the 96 x 96 px [test image](https://user-images.githubusercontent.com/17863119/28234486-5b74ee6c-6900-11e7-9744-84c71140ea12.png) and compare it to the result of my last posting.

Your results probably look bad because you haven't finished the training... it takes a few days on a single GPU. The untrained weights are basically just random values – it can't work that way!

BTW: If you have a problem it's best if you explain what you did so that other people can retrace your steps. How did you test your own image? What weights did you use? Did you finish the training?

**Comment by Andreababy** [comment]
i trained my epoch（1314/2000）.it takes me so much time，so i quit it during the training.i can test the picture from the test dataset，it is very good .i input my own picture as the readme shows.

**Comment by subzerofun** [comment]
These are some samples that i've generated a few minutes ago – please view the png at full size!
The first two look great, but the third has some artifacts in the branches of the trees.

What kind of image were you trying to upscale? Could you upload it?

### Left: Bicubic 4x  |  Middle: SRGAN upscaling  |  Right: Original

![test-collage-1](https://user-images.githubusercontent.com/17863119/28236833-cc20a91e-6930-11e7-8682-01303b175189.png)

**Comment by Andreababy** [comment]
![Uploading 000003.jpg…]()
![Uploading 000004.jpg…]()
![Uploading 000005.jpg…]()

**Comment by jierice** [comment]
@subzerofun @Andreababy Could you guys repost the trained model or send it to me (jasongeophysics@gmail.com)? The link above doesn't seem to work now. Also, can this be run without GPU? Thanks!

**Comment by Andreababy** [comment]
i think it can run with CPU.but it will spend a lot of days. i don't finish the full training.

**Comment by jierice** [comment]
Thanks for your quick reply. Could you please send me the model when you finish? Or do you still have the model from the link above?

**Comment by zsdonghao** [comment]
Hi, the new implementation of [SubpixelConv2d](https://github.com/zsdonghao/tensorlayer/pull/176) solve the compile time problem, you may need to retrain the model if you use the new implementation ..   

You can also use the old implementation via `SubpixelConv2d_old`.

**Comment by ShieLian** [comment]
If anyone doesn't want to retrain your model and also want short compile time, try this brunch:
[patch-2](https://github.com/ShieLian/tensorlayer/tree/patch-2)

**Comment by jiaozixian** [comment]
@subzerofun Could you send me your model? your link shows"transfer deleted". 345239728@qq.com     thanks very much.

**Comment by tuxzz** [comment]
My single GTX780 is hopelessly to reach the end. XD

**Comment by mrveozsahn** [comment]
what s the last total loss values?  I comleted all 2000 epoch but  ı  lost the value before  ı record it

**Comment by JimmyChame** [comment]
Hi, May I have a look at you loss values (including d_loss and g_loss) please if you save them during training? I just want to know what is the normal change of the loss values during training. @subzerofun @zsdonghao

**Comment by yafaraz** [comment]
@subzerofun Sir I need your weights files if u can share them again. The link of wetransfer.com isn't working anymore.
TIA
